Add module variable to allow write permission

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds a module variable (`var.read_only`) that allows for the addition of write permissions to the IAM role being created.

> [!NOTE]
> This change is only possible because of cisagov/s3-read-role-tf-module#47.

It also loosens a version constraint to allow the module to work with a wider range of future versions of the AWS Terraform provider.

## 💭 Motivation and context ##

This change associated with write permissions is required because we require a read-write role that allows access to a particular COOL environment's remote state.

The version constraint change agrees with the advice given [here](https://developer.hashicorp.com/terraform/language/expressions/version-constraints#terraform-core-and-provider-versions).

Note also that these changes have been made in a way that is backward compatible with the role's previous behavior, so these are non-breaking changes.

## 🧪 Testing ##

All automated tests pass.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Revert to the default branch of [cisagov/s3-read-role-tf-module](https://github.com/cisagov/s3-read-role-tf-module) once cisagov/s3-read-role-tf-module#49 is merged.

jsf9k committed:
Add a variable that determines whether the role will be read-only or not

By the default the value if true, indicating that the role should be
read-only.  This agrees with the way the role functioned previously,
so the change should cause no surprises.

If the value is set to false then the role created will allow for
writing in addition to reading.

Note that this new value is passed to cisagov/s3-read-role-tf-module
in read_terraform_state_role.tf.

jsf9k committed:
Add read_only to module outputs

Also update variable and output descriptions to use the word "access"
instead of "read" since writing may be allowed based on the value of
read_only.

jsf9k committed:
Update the role description to specify the type of access

The type of access is either read-only or read-write.

jsf9k committed:
Update module description

Use the term "access" rather than "read" since writing may be allowed
depending on the value of var.read_only.

dav3r reviewed (approved):
This looks great to me.  👍 👍 

My only thought is that with this change, the repo and module should now be renamed to something like `terraform-state-access-role-tf-module`, but I'm ok with just creating an issue for that change since it's a bit of a hassle.

mcdonnnj commented on variables.tf:
```diff
@@ -57,13 +57,19 @@ variable "create_assume_role" {
 variable "role_description" {
-  default     = "Allows read-only access to the Terraform state at '%s' for the '%s' workspace(s) in the %s S3 bucket."
-  description = "The description to associate with the IAM role (as well as the corresponding policy) that allows read-only access to the specified state in the specified S3 bucket where Terraform state is stored.  Note that the first \"%s\" in this value will get replaced with the terraform_state_path variable, the second \"%s\" will get replaced with the terraform_workspace variable, and the th ...
+  default     = "Allows %s access to the Terraform state at '%s' for the '%s' workspace(s) in the %s S3 bucket."
```
I don't see an update to `local.role_description` in this PR that reflects this change.

jsf9k replied:
Excellent catch!  See commit 52075df.

jsf9k committed:
Actually update the role description

I neglected to do this, but fortunately @mcdonnnj caught my error.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
> This looks great to me. 👍 👍
> 
> My only thought is that with this change, the repo and module should now be renamed to something like `terraform-state-access-role-tf-module`, but I'm ok with just creating an issue for that change since it's a bit of a hassle.

Please see issue #38 (and cisagov/s3-read-role-tf-module#48).

mcdonnnj reviewed (approved):
LGTM ✔

dav3r commented:
> Please see issue #38 (and [cisagov/s3-read-role-tf-module#48](https://github.com/cisagov/s3-read-role-tf-module/issues/48)).

Perfect, thanks! 😎

jsf9k committed:
Loosen version constraint to allow more future versions

This agrees with the advice given here:
https://developer.hashicorp.com/terraform/language/expressions/version-constraints#terraform-core-and-provider-versions

jsf9k committed:
Temporarily use a non-default version of cisagov/s3-read-role-tf-module

This is being done for testing purposes and can be reverted once
cisagov/s3-read-role-tf-module#49 is approved and merged.

jsf9k committed:
Revert to the default branch of cisagov/s3-read-role-tf-module

We can do this now that cisagov/s3-read-role-tf-module#49 her been
merged.
